### PR TITLE
[ENH] Rework of base series annotator API

### PR DESCRIPTION
**Pull request by Alex-JG3**
#### Reference Issues/PRs

See #3214 .

#### What does this implement/fix? Explain your changes.

Refactors the `BaseSeriesAnnotator` class:

* Removes the `fmt` and `label` attributes.
* Adds the `learning_type` and `task` attributes.
* Adds default `predict` and `transform` method.
* Adds default `predict_points` and `predict_segments` methods.
* Adds methods for converting  between dense and sparse output formats.
* Adds default methods for `predict` and `transform`.

#### Does your contribution introduce a new dependency? If yes, which one?

No.

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [ ] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

File: sktime/annotation/adapters/_pyod.py
Comment (fkiraly) [inline sktime/annotation/adapters/_pyod.py:30]: isn't `pyod` always unsupervised?

**Comment by fkiraly** [comment]
As discussed in the last developer meeting, the approach using tags is showcased here: https://github.com/sktime/sktime/pull/6271

**Comment by fkiraly** [comment]
PS: if you migrate to tags, you need to merge the parts of https://github.com/sktime/sktime/pull/6271 as well that add tags to the registry.

**Comment by Alex-JG3** [comment]
@fkiraly, @duydl I am at a small blocker and would appreciate your feedback.

### Converting between sparse and dense formats.

I have added the static methods `sparse_to_dense` and `dense_to_sparse` for converting between sparse and dense formats. I think having these methods is important because:

* `sparse_to_dense` can be used to convert the output from `predict_segments` or `predict_change_points` to the dense format.
* Suppose we have a segmentation annotator. One can get the segments in dense format by using `transform` but one needs `sparse_to_dense` to get change points (from `predict_change_points`) in dense format.

Here is the blocker. Suppose we have an array of anomalies in sparse format `[2, 5, 7]`. If we use `sparse_to_dense` on this array, we get,

```python
>>> import numpy as np
>>> from sktime.annotation.base._base import BaseSeriesAnnotator
>>> y_sparse = np.array([2, 5, 7])  # Indices of changepoints/anomalies
>>> BaseSeriesAnnotator.sparse_to_dense(y_sparse)
array([0, 0, 1, 0, 0, 1, 0, 1], dtype=int32)
```

This assumes that input `X` used to generate the array of anomalies contains 8 samples. This assumption is incorrect. To solve this, we need to pass the length of `X` to `sparse_to_dense` so we can pad the output with 0's such that it has the same length as `X`; this feels a little clumsy to me. We could make `dense_to_sparse` a private method that is only called from `transform` but I think `dense_to_sparse`/`sparse_to_dense` are handy bits of functionality that would be shame to lose.

* Do you think `dense_to_sparse`/`sparse_to_dense` would be useful to provide to users?
* Do you think passing the length of `X` to `sparse_to_dense` feels clumsy? If so can you think of an alternative?

**Comment by fkiraly** [comment]
Interesting. My opinion below.

### on the design

I think the key observation is that the functions `sparse_to_dense` and `dense_to_sparse` are meant to be inverse to each other. However, mathematically, from an "abstract data type" perspective, this is only true, and can only be true, given knowledge of the index of `X`.

In the "numpy" situation, the index of `X` is an abstract `RangeIndex`, and it is entirely determined by its length. Therefore, including length does not feel clumsy to me, but mathematically necessary.

### as public function

I also think it is beneficial to have these functions, at least we shouldn't remove them as some users might have been using them. I would add a length parameter, with a default that leads to current behaviour.

I would also leave them external, since very long-term, I think this should live in an abstract conversion module similar to `datatypes.convert` and adjacent functionality, and leaving them loose is closer to centralizing this functionaliy.

A similar case would be converting between different representations of segmentations.

**Comment by duydl** [comment]
> Do you think passing the length of X to sparse_to_dense feels clumsy? If so can you think of an alternative?

The change points outputs of ruptures library are sparse but include X length at the last index. This could be an alternative that makes sparse more equivalent to dense format. It might also be a little confusing though.

> Do you think dense_to_sparse/sparse_to_dense would be useful to provide to users?

@fkiraly I think these are new, so we don't have to worry about legacy problems. 

The functions would allow users to convert outputs to other formats directly. Do we still need the other workflow then; calling class methods again on the same input to get different formats would require either caching input and output or refitting, at least for the current estimators?

btw, great work on the converting methods. I have been wondering how to distinguish between dense format of two-class segmentation and change points. Some further additions I could think of:
* function converting from scores to labels
* check if y is not float in dense to sparse

**Comment by fkiraly** [comment]
Interesting analysis, @duydl.

> The change points outputs of ruptures library are sparse but include X length at the last index. This could be an alternative that makes sparse more equivalent to dense format. It might also be a little confusing though.

Ouch - I agree with the "confusing" remark, this is basically a "data type representation hack", I would not recommend that.

I would take this as a validation of the observation that the information on index (including length) is crucial.

> @fkiraly I think these are new, so we don't have to worry about legacy problems.

Are they? Sorry, I haven't been monitoring the module closely.

If they are, how about we hide them in a private module for the time being, but keep them external to the class:

* private module gives us more flexibility to make breaking changes as wel consolidate the interface
* class external gives us more flexibility to move the eventual location, and avoid "mudballing" around these


Minor further points raised:

> function converting from scores to labels

Will this not be algorithm specific?

> i don't think sparse change point is equivalent to sparse segment i.e convertible

It is not equivalent, but there's a "canonical" one-way conversion, I thought?
Please correct me if you disagree.

**Comment by duydl** [comment]
> It is not equivalent, but there's a "canonical" one-way conversion, I thought?

Sorry, it was a little out of context. I am referencing to the code, there is a `change_points_to_segments(y_sparse):` that I don't think is feasible.

**Comment by fkiraly** [comment]
@Alex-JG3, can you kindly give an update on timelines? I would like to get a working version in before the workstreams kick off May 1-7, to enable refactoring and extension.

Let me know if you don't find time, then I'll complete it.

**Comment by Alex-JG3** [comment]
Re the length issues, I have gone for adding it as an argument to `sparse_to_dense`.

> Are they? Sorry, I haven't been monitoring the module closely.

Yes they are new. Other annotators implement their own versions of these functions but there is no consistent way of doing these conversions.

@duydl, regarding your suggestions.

> function converting from scores to labels

I have left out the scoring for the time being. I think this stuff should go in another PR.

> check if y is not float in dense to sparse

Yes, I agree. I would like to leave this to a future PR.

> I think Dense scores should also be a possible output of transform.

I would prefer something like `score_samples` a bit like Isolation forests in sklearn. The output would be dense. https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.IsolationForest.html. Again, I would prefer to leave this for another PR.

## Timelines

Here is where I think this PR is for the changes listed at the top.

[x] Removes the `fmt` and `label` attributes.
[x] Adds the `learning_type` and `task` attributes.
[x] Adds default `predict_points` and `predict_segments` methods.
[x] Adds methods for converting between dense and sparse output formats.
[ ] Adds default methods for predict and transform

For the last one, there is default transform method but not a default predict method. I think a default predict method should be added when we start refactoring some of the annotators to use `predict_segments`/`predict_points`. All of the current annotators use `predict` so I think it would break too many of them to make the change here.

In terms of changes to `BaseSeriesAnnotator`, I think this is a natural place to stop for the time being. Please let me know if there are other things you think still need to be added.

## Extras things that should be added to this PR

1. Add new tags to the registry.
2. Unit tests for `BaseSeriesAnnotator`.
3. Update one of the algorithms (probably ClaSP) to use some of the new features of the `BaseSeriesAnnotator`.
4. Put `sparse_to_dense` and `dense_to_sparse` into their own private module.

@fkiraly, do you agree with this list? I think it would be easier if you, @fkiraly, handle 1 since I am not sure what you mean by this. For 2, I think the main methods to test for the time being are `dense_to_sparse` and `sparse_to_dense`. For 3, again I am happy to take a stab at this. For 4 , @fkiraly could you please take a look at this since I am not sure exactly what you mean? 

I can work on this tomorrow evening, Saturday and Sunday.

Also it would be good to get a review.

**Comment by fkiraly** [comment]
Thanks for the update! I think this is indeed close to finish, and the next steps as suggested make a lot of sense.

To answer in more detail:

>  I think it would be easier if you, @fkiraly, handle 1 since I am not sure what you mean by this

I think the tags have already been added, in `registry._tags` - a while back you asked the same already, so it's done.

> For 4 , @fkiraly could you please take a look at this since I am not sure exactly what you mean?

I don't think this is strictly necessary - I just meant, moving it to a different file that starts with an underscore, and importing from there, just to keep the separation. Not a blocker imo though.

File: sktime/annotation/base/_base.py
Comment (fkiraly) [inline sktime/annotation/base/_base.py:105]: if these are tags, they are not set by the user.
I think we could move this to the test suite, hence. But it also does not harm here, for now.

File: sktime/annotation/base/_base.py
Comment (fkiraly) [inline sktime/annotation/base/_base.py:360]: this should clarify whether you mean `loc` or `iloc`.

Further, I think it needs to be able to represent ends as well.

The format I've used in private code is `pd.DataFrame`, with columns `seg_start` and `seg_end`, and optional `seg_label`.

Noting that, to compress, you could remove:

* `seg_end`, if segments are consecutive and exhaustive
* `seg_label`, if segments are unlabelled

There is no way I see in the `np.ndarray` case to map the four cases you can have (having or not having either of the two) on different shapes (there are only three cases if you have 3 cols max).

File: sktime/annotation/base/_base.py
Comment (fkiraly) [inline sktime/annotation/base/_base.py:387]: as above, I feel we should account for the option to return labels too.

I have a small preference for `pd.DataFrame`, by reasoning of consistency.

File: sktime/annotation/base/_base.py
Comment (fkiraly) [inline sktime/annotation/base/_base.py:395]: should this call the public or the private method, i.e., `predict_segments` or `_predict_segments`?
In many similar cases, we call the public method, since that has the call contract guaranteed.

Comment (Alex-JG3) [reply sktime/annotation/base/_base.py:392]: Yes it should call `predict_segments` not `_predict_segments`, I will update.

File: sktime/annotation/base/_base.py
Comment (Alex-JG3) [reply sktime/annotation/base/_base.py]: Are you okay with the following?
* dense segments being a series amd sparse segments being a dataframe.
* sparse and dense change points being a series.
* sparse and dense anomalies being a series.

Using series avoids having to have different column names for segmentation, anomaly detection, and change point detection. I think it makes the code cleaner.

> this should clarify whether you mean loc or iloc

I haven't included this in the docstrings yet but I would prefer `iloc` for the time being. I think unordered indexes could cause issues `loc` and sparse segmentation. I think we could get around this with some proper type checking but in the interest of time, I would rather leave this for the time being.

File: sktime/annotation/base/_base.py
Comment (Alex-JG3) [reply sktime/annotation/base/_base.py]: Okay, I have had a go at making the switch to pandas dataframes and series.

File: sktime/annotation/base/_base.py
Comment (fkiraly) [reply sktime/annotation/base/_base.py]: > Using series avoids having to have different column names for segmentation, anomaly detection, and change point detection. I think it makes the code cleaner.

hmm - how do you represent segments that are not "tiled", i.e., the union is not the entire range, or they do overlap?

Example:
* original time series (to segment) has indices 1, 2, 3, 4, 5
* segments go [1, 2] and [4, 5]

Example 2:
* same time series
* segments go [1, 3] and [2, 4]

Comment (fkiraly) [reply sktime/annotation/base/_base.py]: > I think unordered indexes could cause issues `loc` and sparse segmentation.

So, points "between indices" is not allowed?

Why this could make sense: let's say you are using the anomaly detection metric which is the distance to the closest anomaly plus a cost per anomaly that you detect.

This is in general optimized by predicting it to be somewhere in-between the indices, rather than restricting it to observed indices only.

Comment (fkiraly) [reply sktime/annotation/base/_base.py]: (feel free to ignore any of the above, if it leads to "unclear design" issues)

Comment (Alex-JG3) [reply sktime/annotation/base/_base.py]: Thanks for raising these points.

> the union is not the entire range.

In sparse format I don't think that is too much of problem. Indexes that are not covered by the segments are unclassified.

In dense format, I suggest we reserve an integer to represent unclassified segments. I think `-1` would make the most sense since `0` is reserved is used to represent none changepoints/anomalies.

Here, I check for zeroes. https://github.com/Alex-JG3/sktime/blob/e7a748671488a7d945bc8a2095fb4bf9e6edc125/sktime/annotation/base/_base.py#L538.

> or they do overlap?

Currently, it is assumed the segments do not overlap. However, in sparse format I don't think is too difficult to represent. Your example would be dataframe that looks like (I have taken 1 of the indexes to account for iloc instead of loc):

```python
   seg_label  seg_start  seg_end
0          1          0        1
1          2          2        3
```

Converting this into dense is more difficult. I would suggest we use a dataframe where each column represents a different label. Your example becomes:

```python
   label_1  label_2
0        1        0
1        1        0
2        0        0
3        0        1
4        0        1
```

It would be nice deal with this in a way without so many zeros. But, I don't think lists in cells is well supported in pandas.

> So, points "between indices" is not allowed?

Not currently but I see your point.

This sounds especially relevent if the timeseries is not uniformly sampled. For anomalies in between indexes, I would say we need `loc` instead of `iloc`. I think this particular point adds quite a bit of complexity that would be better handled carefully in its own PR. I think the other points can be dealt with now.

Comment (fkiraly) [reply sktime/annotation/base/_base.py]: > Your example would be dataframe that looks like (I have taken 1 of the indexes to account for iloc instead of loc):

Yes, that makes sense. `seg_end` could be optional and default to the `seg_start` of the next row - though in this case one would have to worry about the end of the last one. So not sure if this is fully thought out.

> Converting this into dense is more difficult. I would suggest we use a dataframe where each column represents a different label. Your example becomes:

Hm, that's one-hot encoded.
As long as they do not overlap, you could still have unique label, and 0 or -1 for no label?

There's pros/cons for adopting a 1D representation as long as it is possible. Con is that you have to switch representations if not; Pro is that 1D is useable more universally (e.g., as a target with an estimator that needs 1D `y`)

> I think this particular point adds quite a bit of complexity that would be better handled carefully in its own PR.

Agreed.

Comment (fkiraly) [reply sktime/annotation/base/_base.py]: PS: have you seen `pandas.IntervalIndex`?

It can be used as a column type:
https://pandas.pydata.org/docs/reference/api/pandas.IntervalIndex.html

Not arguing that we use it, as it may come with its own issues.

There is an interesting parallel, the `from_breaks` method, related to the discussion above.

Comment (Alex-JG3) [reply sktime/annotation/base/_base.py]: > have you seen `pandas.IntervalIndex`?

I havn't is that `is_overlapping` function looks like it would be useful.

**Comment by Alex-JG3** [comment]
@fkiraly, I have had a first rough attempt at refactoring `ClaSPSegmentation`. There are a few of things that would be good to get your feedback on:

* Despite the class name this algorithm seems to be closer to a change point detection algorithm than a segmentation algorithm. I have thus switched its `task` to `change_point_detection`.
* Since I have not changed `predict` in `BaseSeriesAnnotator` yet (since it will break a lot of the other annotators), I have had to implement `_predict` in `ClaSPSegmentation`.
* The `_predict_scores` method relied on `self.fmt` so I have changed it to return a tuple of scores and profiles for the time being.

**Comment by fkiraly** [comment]
> Despite the class name this algorithm seems to be closer to a change point detection algorithm than a segmentation algorithm. 

Yes, I agree, especially due to the sliding window approach applied internally.

It seems explicitly misclassified in the original publication as well - which is not unusual though, as there is a close correspondence  due to the 1:1 reduction mapping via "segments are the intervals between two adjacent changepoints".

Though the difference is also very subtle - if within-segment behaviour is qualitatively stationary, there is no generative difference between the two.

But, long story short, I would indeed classify it as changepoint detection algorithm.

> I have had to implement `_predict` in `ClaSPSegmentation`.
> The `_predict_scores` method relied on self.fmt so I have changed it to return a tuple of scores and profiles for the time being.

This makes sense to me, I think it is directly implied by the target interface definition.

**Comment by fkiraly** [comment]
PS: apologies for killing your CI multiple times - just giving the 0.29.0 release the right of way...

**Comment by Alex-JG3** [comment]
@fkiraly, the pipeline is failing due to a dependency that cannot be installed. Could you help with this? I am not sure what the problem is.

**Comment by fkiraly** [comment]
I cannot see the logs due to this issue with the logs: https://github.com/sktime/skpro/issues/304

However, if it is something about a SHA not matching, restarting the failed jobs should fix it, it is sporadic.

**Comment by fkiraly** [comment]
@benHeid or @yarnabrina, can you see the logs?

**Comment by yarnabrina** [comment]
Thus was the error

```
ERROR: THESE PACKAGES DO NOT MATCH THE HASHES FROM THE REQUIREMENTS FILE. If you have updated the package versions, please update the hashes. Otherwise, examine the package contents carefully; someone may have tampered with them.
    unknown package:
        Expected sha256 55b380cff92d493e465b23e96a4bce8eec1a7b02db92c7718bb0d883f244c2b9
             Got        f8a96f8553d80288955fd9fa54bb4f9e6aec5df2028074a8cc65a11510af611a
```

I restarted that job, hopefully it'd work this time. 

(Your logs issue by AVG is very weird, we need to find a way around soon I think)

**Comment by fkiraly** [comment]
Tests pass - is this ready for a "first slice merge"?

**Comment by Alex-JG3** [comment]
> tests pass - is this ready for a "first slice merge"?

I would like to check this over one final time before merging. I have switched over to using `pandas.IntervalIndex` meaning I have also switched over from using an `iloc` style to a `loc` style like we discussed earlier. I think this will be better in the end but there are still a few rough edges.

I have time to work on this again on Thursday.

**Comment by Alex-JG3** [comment]
I am marking this as ready for review but there is one thing I haven't quite been able to find a good solution for. I want `dense_to_sparse` and `sparse_to_dense` to be opposites of each other.

Suppose we have the following dense segments.

```
[1, 1, 2, 2, 3, 3]
```

The sparse format of this would be,

```
[0, 2)  1
[2, 4)  2
[4, 5)  3
```

Converting this back into dense would give us,

```
[1, 1, 2, 2, 3, -1]
```

From the sparse representation, index 5 is not included in the interval for the third segment hence it is labelled -1. Here are a few options I considered for dealing with this:

## Set the final end point to infinity

If we have dense segments like,

```
[1, 1, 2, 2, 3, 3]
```

We cannot say when the third segment ends so we should assume it never ends. So the dense form would look like,

```
[0, 2)    1
[2, 4)    2
[4, Inf)  3
```

This might be okay for integers/floats but having infinite datetimes is nasty.

## Include the right limit of the final segment

Suppose again we have dense segments like,

```
[1, 1, 2, 2, 3, 3]
```

And in sparse format,

```
[0, 2)  1
[2, 4)  2
[4, 5)  3
```

When we convert back to dense, we could just say index 5 is labelled 3. This felt like it was going against the information in the sparse format and potentially confusing.

## Use closed intervals

I have opted for using left closed intervals. We could use left and right closed intervals like,

```
[0, 2]  1
[2, 4]  2
[4, 5]  3
```

But now index 2 is labelled both 1 and 2. I think using right closed intervals is less ambiguous. Unfortunately you don't seem to able to use a mix of left and right closed intervals.

The current solution feels like the best of a bad bunch. Please let me know what you think or if you have any other solutions.

**Comment by fkiraly** [comment]
I am currently reviewing and will write a longer review.

Regarding the last question, would a solution not listed clear things up: transforming the last interval to `[4, 6)` ?

**Comment by Alex-JG3** [comment]
> I am currently reviewing and will write a longer review.
> 
> Regarding the last question, would a solution not listed clear things up: transforming the last interval to `[4, 6)` ?

For that example yes but I think it becomes less clear if we have an index that is not equally spaces. For example, taking the example from before and making it into a proper pandas series we get,

```
0 1
1 1
2 2
3 2
4 3
5 3
```

the interval of the final segment could be `[4, 6)` and that would be sensible. But suppose the index is not equally spaces.

```
0.1 1
1.1 1
2.2 2
2.3 2
4.1 3
4.3 3
```

For example, I don’t think an interval of `[4.1, 5.1)` would be sensible for the third segment. For completeness, the sparse format of the series above would in the current implementation would be,

```
[0.1, 2.2) 1
[2.2, 4.1) 2
[4.1, 4.3) 3
```

**Review by fkiraly — APPROVED** [review]
Very nice!

I think this will be the foundation of finally sorting this module out.
I also like how this removes boilerplate from the concrete classes.

The one worry I have is about removal of arguments and that the changes here are not deprecation safe.
The module is clearly marked as "experimental", so we could do that, I wonder though whether we can do sth, like leaving the old args, and raising warnings that they will disappear, instead.

**Comment by fkiraly** [comment]
> For that example yes but I think it becomes less clear if we have an index that is not equally spaces. 

I see. How about making the last interval run over by the last difference, then?
In the unequally spaced example, [4.1, 4.) then.

I would also like to summon @vascosch92 who might have a useful opinion here, given sequences are involved.

The greater question is coming up with mutually inverse functions from a value representation to a segment representation, as above. This is not possible if the target index is not known (only partial inverses), but the function assume we know `X.index` - see above, discussion starts here in a self-contained form: https://github.com/sktime/sktime/pull/6265#issuecomment-2118828155

**Comment by VascoSch92** [comment]
> I would also like to summon @VascoSch92 who might have a useful opinion here, given both sequences and sets are involved at the same time - hope you enjoy this 😁
:-) 

Just some thoughts from my side:

I think the solution of @Alex-JG3 is good. What are you doing here is to represent in an intelligent way the values of a column/serie, i.e., you want to group interval of **contiguous** indexes where the column/serie is constant.

After that if you want to consider closed/open/right-closed/... interval, it is just taste, but it should be written clearly in the documentation or/and in the docstring.

The choice of taking left-closed and right-open interval ( for example  `[4.1, 4.5)` ) seems legit, as it is also what Python do with lists and slicing, i.e., a list start from index `0` and ends at index `length-1`.

> For that example yes but I think it becomes less clear if we have an index that is not equally spaces.
I see your point but 
- the index of a `df` is always monotonically increasing (is correct? I'm not sure now)
- you said that you consider interval of contiguous indices with the same value.
Therefore, it should be clear.

**Comment by fkiraly** [comment]
> the index of a df is always monotonically increasing (is correct? I'm not sure now)

Not for `pandas.DataFrame` in general, but we can enforce this in our output assumptions (should be tested, let's make a note of this!)

**Comment by Alex-JG3** [comment]
> the index of a df is always monotonically increasing (is correct? I'm not sure now)

No the index is not currently always monotonically increasing but it should be. In fact I think some of the new methods will break if index is not monotonically increasing. We need to enforce this.

I have added deprecation warnings in clasp and the pyod annotator. I think those are the only two classes that are affected by deprecation. Currently, the deprecation message feels very brief:

```
The fmt argument is going to be removed.
```

Is this sufficient? Do we want to point a specific version when these arguments will be depracted?

**Comment by fkiraly** [comment]
> Is this sufficient? Do we want to point a specific version when these arguments will be depracted?

I would say sth like

`f"Warning from {type(self).__name__}: fmt argument will be removed in 0.31.0. For behaviour equivalent to current fmt=a usage, do X; for fmt=b usage, do Y"`

**Comment by Alex-JG3** [comment]
I have rewritten the deprecation warning. I have also removed the deprecation warning for `labels` for the pyod adapter since the changes to `BaseSeriesAnnotator` do not deal with scores properly yet. This will come in a future PR.

If everyone else is happy are we ready to merge?

**Comment by Alex-JG3** [comment]
Could someone please rerun the CI/CD pipeline? Looks like the docs jobs has failed due to running out of time to build. I don't think this as a result of changes introduced in this PR.

**Comment by fkiraly** [comment]
Restarted - I also updated the deprecation messages and added release manager notes.

Question, it seems we want to remove the `labels` param in `PyODAnnotator` too - should there be a deprecation warning as well?

**Review by fkiraly — APPROVED** [review]
I think we are ready to merge for release with 0.30.0.

@Alex-JG3, it would be great if you could check the updated warnings and release manager comments.

**Comment by fkiraly** [comment]
ok, let´s leave `indicator` in for now, then.

**Pull request merged**
